src/docfx/build/moniker/MonikerProvider.cs

```diff
@@ -81,36 +83,57 @@ public SourceInfo<string> GetConfigMonikerRange(FilePath file)
             return (null, monikers);
         }
 
-        private (Error? error, string[] monikers) GetFileLevelMonikersCore(FilePath file)
+        private (List<Error> errors, string[] monikers) GetFileLevelMonikersCore(FilePath file)
         {
+            var errors = new List<Error>();
             var (_, metadata) = _metadataProvider.GetMetadata(file);
 
             var configMonikerRange = GetConfigMonikerRange(file);
-            var configMonikers = _rangeParser.Parse(configMonikerRange);
+            var configedMonikers = _rangeParser.Parse(configMonikerRange);
 
-            if (!string.IsNullOrEmpty(metadata.MonikerRange))
+            if (!string.IsNullOrEmpty(metadata.MonikerRange) || metadata.Monikers.Length > 0)
             {
                 // Moniker range not defined in docfx.yml/docfx.json,
                 // user should not define it in file metadata
-                if (configMonikers.Length == 0)
+                if (configedMonikers.Length == 0)
                 {
-                    return (Errors.Versioning.MonikerRangeUndefined(metadata.MonikerRange), configMonikers);
+                    var source = metadata.MonikerRange.Source;
+                    if (source != null)
+                    {
+                        errors.Add(Errors.Versioning.MonikerRangeUndefined(source));
+                    }
+                    return (errors, configedMonikers);
+                }
+
+                // monikerRagen takes precedence over monikers since it is more likely from user configuration
+                if (!string.IsNullOrEmpty(metadata.MonikerRange) && metadata.Monikers.Length > 0)
+                {
+                    var source = metadata.Monikers.First().Source;
+                    if (source != null)
+                    {
+                        errors.Add(Errors.Versioning.DuplciateMonikerConfig(source));
+                    }
                 }
 
                 var fileMonikers = _rangeParser.Parse(metadata.MonikerRange);
-                var intersection = configMonikers.Intersect(fileMonikers).ToArray();
+                if (string.IsNullOrEmpty(metadata.MonikerRange) && metadata.Monikers.Length > 0)
+                {
+                    fileMonikers = metadata.Monikers.SelectMany(x => _rangeParser.Parse(x)).ToArray();
```

Comment:
consider do validation only

Reply:
Since we do not have a place to easily get `MonikerDefinition`, which is the valid moniker list.
It only exists within [EvaluatorWithMonikersVisitor](https://github.com/dotnet/docfx/blob/cff6a06bbc9c9ae6a8a08f5bc96e591fe0acc1ab/src/docfx/build/moniker/MonikerProvider.cs#L39), then ties to `MonikerRangeParser`.
I would leave this logic as it is since we need to refactor the whole moniker parsing logic quite a bit to split the validation only logic out. I suspect that we would have that much side effect of reusing the `MonikerRangeParser`